ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+  FROM http_request
+  GROUP BY zone_id, minute;
+
+Now dashboard queries are a little more complicated, you have to read out the cardinality
+during SELECT:
+
+.. code-block:: sql
+
+  SELECT
+    request_count, success_count, error_count, average_response_time_msec,
+    hll_cardinality(distinct_sessions) AS distinct_session_count
+  FROM http_request_1min
+  WHERE zone_id = 1 AND minute = date_trunc('minute', now());
+
+HLLs aren't just faster, they let you do things you couldn't previously. Say we did our
+rollups, but instead of using HLLs we saved the exact unique counts. This works fine, but
+you can't answer queries such as "how many distinct sessions were there during this
+one-week period in the past we've thrown away the raw data for?". With HLLs, it's easy:
+
+.. code-block:: sql
+
+  -- careful, doesn't work!
+  SELECT
+    hll_cardinality(hll_union_agg(distinct_sessions))
+  FROM http_request_1day
+  WHERE ingest_time BETWEEN timestamp '06-01-2016' AND '06-08-2016';
+
+Well, it would be easy, except since Citus `can't yet
+<https://github.com/citusdata/citus/issues/120>`_ push down aggregates such as
+hll_union_agg. Instead you have to do a bit of trickery:
+
+.. code-block:: sql
+
+  -- this should be run on the workers and master
+  CREATE AGGREGATE sum (hll)
+  (
+    sfunc = hll_union_agg,
+    stype = internal,
+  );
+
+Now, when we call SUM over a collection of hlls, postgresql will return the hll for us.
+This lets us write the above query as:
+
+.. code-block:: sql
+
+  -- working version of the above query
+  SELECT
+    hll_cardinality(SUM(distinct_sessions))
+  FROM http_request_1day
+  WHERE ingest_time BETWEEN timestamp '06-01-2016' AND '06-08-2016';
+
+More information on HLLs can be found in `their github repo
+<https://github.com/aggregateknowledge/postgresql-hll>`_.
+
+HLLS are an example from the postgresql community, but there are a couple extensions we've
+written ourselves which do the same thing (improve performance and storage requirements)
+for different kinds of queries. This includes `count-min sketch
+<https://github.com/citusdata/cms_topn>`_ for
+top-n queries, and `HDR <https://github.com/citusdata/HDR>`_, for percentile queries.
+
+Unstructured Data with JSONB
+----------------------------
+
+Citus works well with Postgres' built-in support for unstructured data types. To
+demonstrate this, let's keep track of the number of visitors which came from each country.
+Using a semi-structure data type saves you from needing to add a column for every
+individual country and blowing up your row width.  We have `a blog post
+<https://www.citusdata.com/blog/2016/07/14/choosing-nosql-hstore-json-jsonb/>`_ explaining
+which format to use for your semi-structured data. It says you should usually use jsonb
+but never says how. Let's correct that :)
+
+First, add the new column to our rollup table:
+
+.. code-block:: sql
+
+  ALTER TABLE http_requests_1min ADD COLUMN country_counters (JSONB);
```

Comment:
I think counters is not a great use of JSONB, since the entire row is re-written whenever you update a counter. Is it possible to add something to the http_request table? e.g. cookies?